src/dex/gmx/usdg.ts

```diff
@@ -0,0 +1,82 @@
+import { Interface } from '@ethersproject/abi';
+import { DeepReadonly } from 'ts-essentials';
+import { PartialEventSubscriber } from '../../composed-event-subscriber';
+import {
+  Address,
+  MultiCallInput,
+  MultiCallOutput,
+  Logger,
+  Log,
+  BlockHeader,
+} from '../../types';
+import { USDGState } from './types';
+import ERC20ABI from '../../abi/erc20.json';
+import { Lens } from '../../lens';
+import { NULL_ADDRESS } from '../../constants';
+
+export class USDG<State> extends PartialEventSubscriber<State, USDGState> {
+  static readonly interface = new Interface(ERC20ABI);
+
+  constructor(
+    private usdgAddress: Address,
+    lens: Lens<DeepReadonly<State>, DeepReadonly<USDGState>>,
+    logger: Logger,
+  ) {
+    super([usdgAddress], lens, logger);
+  }
+
+  getTotalSupply(state: DeepReadonly<State>) {
+    return this.lens.get()(state).totalSupply;
+  }
+
+  public processLog(
+    state: DeepReadonly<USDGState>,
+    log: Readonly<Log>,
+    blockHeader: Readonly<BlockHeader>,
+  ): DeepReadonly<USDGState> | null {
+    try {
+      const parsed = USDG.interface.parseLog(log);
+      const _state: USDGState = state;
```

Comment:
Same as other comment, you are modifying the state in the state history, you need to copy this or construct a new object.